pkg/controller/container-runtime-config/container_runtime_config_controller.go

```diff
@@ -1026,41 +1066,101 @@ func registriesConfigIgnition(templateDir string, controllerConfig *mcfgv1.Contr
 	return &registriesIgn, nil
 }
 
-// getValidScopePolicies returns a map[scope]policyRequirement from ClusterImagePolicy
-func getValidScopePolicies(clusterImagePolicies []*apicfgv1alpha1.ClusterImagePolicy) (map[string]signature.PolicyRequirements, error) {
+// getValidScopePolicies returns a map[scope]policyRequirement from ClusterImagePolicy, a map[scope][namespace]policyRequirement from ImagePolicy CRs.
+// It skips ImagePolicy scopes that conflict with ClusterImagePolicy scopes and logs the conflicting scopes in the ImagePolicy Status.
+func getValidScopePolicies(clusterImagePolicies []*apicfgv1alpha1.ClusterImagePolicy, imagePolicies []*apicfgv1alpha1.ImagePolicy, ctrl *Controller) (map[string]signature.PolicyRequirements, map[string]map[string]signature.PolicyRequirements, error) {
 	clusterScopePolicies := make(map[string]signature.PolicyRequirements)
+	namespacePolicies := make(map[string]map[string]signature.PolicyRequirements)
+
 	for _, clusterImagePolicy := range clusterImagePolicies {
 		sigstoreSignedPolicyItem, err := policyItemFromSpec(clusterImagePolicy.Spec.Policy)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 		for _, scope := range clusterImagePolicy.Spec.Scopes {
 			scopeStr := string(scope)
 			clusterScopePolicies[scopeStr] = append(clusterScopePolicies[scopeStr], sigstoreSignedPolicyItem)
 		}
 	}
 
-	return clusterScopePolicies, nil
+	for _, imagePolicy := range imagePolicies {
+		namespace := imagePolicy.ObjectMeta.Namespace
+		sigstoreSignedPolicyItem, err := policyItemFromSpec(imagePolicy.Spec.Policy)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		var conflictScopes []string
+		for _, scope := range imagePolicy.Spec.Scopes {
+			scopeStr := string(scope)
+			nestedInClusterScope := false
+			for clusterScope := range clusterScopePolicies {
+				if runtimeutils.ScopeIsNestedInsideScope(scopeStr, clusterScope) {
+					conflictScopes = append(conflictScopes, scopeStr)
+					nestedInClusterScope = true
+					break
+				}
+			}
+			if nestedInClusterScope {
+				continue
+			}
+			if _, ok := namespacePolicies[scopeStr]; !ok {
+				namespacePolicies[scopeStr] = make(map[string]signature.PolicyRequirements)
+			}
+			namespacePolicies[scopeStr][namespace] = append(namespacePolicies[scopeStr][namespace], sigstoreSignedPolicyItem)
+		}
+		if ctrl != nil {
+			if len(conflictScopes) > 0 {
+				msg := fmt.Sprintf("has conflict scope(s) %q that equal to or nest inside existing clusterimagepolicy, only policy from clusterimagepolicy scope(s) will be applied", conflictScopes)
```

Comment:
I don't see this constraint being called out in [the API Godocs][1].  But unless it's a short term limitation we hope to address soon, I think we should update those Godocs to explain this interaction, to avoid surprising folks after they've created their ImagePolicy.

[1]: https://github.com/openshift/api/blob/2669d1ebc91092ec32e77b1d3f503b11f759b855/config/v1alpha1/types_image_policy.go#L35-L48

Reply:
Just for the record, this is consistent with the conflict / multi-value handling rules in https://github.com/openshift/enhancements/blob/master/enhancements/api-review/add-ClusterImagePolicy-and-ImagePolicy-for-Signature-Verification.md#update-container-runtime-config-controller-to-watch-clusterimagepolicy-and-imagepolicy .

IIRC it’s been fairly extensively discussed when written, but I didn’t now go back to reread that — I’m just saying that this is an intentional behavior, I’m not saying that it’s not open to change.

Reply:
Having ClusterImagePolicies outweigh ImagePolicies sounds reasonable to me.  I'm just suggesting that the API Godocs (which will eventually be rendered into [API reference docs][1] in whatever 4.y they go generally available) get updated to explain this behavior, so it's less surprising to customer using the Kube types.

[1]: https://docs.openshift.com/container-platform/4.16/rest_api/config_apis/config-apis-index.html

Reply:
I updated the godoc comment in PR  https://github.com/openshift/api/pull/2036